### ISSUE TITLE
Product Import -> Missing categories entry in importfile removes url rewrites

### ISSUE THREAD

**Issue by ulftietze** [magento/magento2#18269]
### Preconditions

1. Magento 2.2.X | probably 2.3 as well
2. PHP 7.0 | 7.1
3. Would be nice to have 2 or more websites

### Steps to reproduce

1. Shop contains an existing product with at least following informations:

sku | categories | url_key | product_websites
------------ | ------------- | ------------- | ------------- 
123456789 | Default Category/ExistingCategory1,Default Category/ExistingCategory2 | custom-url-key-for-product-123456789 | de,at

Attributes like: `name | price | url_key | attribute_set_code`  doesnt matter here.
Actually only interesting point here is: `product_websites`, but im going to explain later

2. URL Rewrite contains following informations, because first time everything was imported successfully and completely:

| url_rewrite_id | entity_type | entity_id | request_path | target_path | redirect_type | store_id | description | is_autogenerated | metadata
------------ | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- 
1 | product |  1 | custom-url-key-for-product-123456789/ | catalog/product/view/id/1 | 0 | 1 | NULL | 1 | NULL 
2 | product |  1 | custom-url-key-for-product-123456789/ | catalog/product/view/id/1 | 0 | 2 | NULL | 1 | NULL                  
3 | product | 1 | existingcategory1/custom-url-key-for-product-123456789/ | catalog/product/view/id/1/category/1 | 0 | 1 | NULL | 1 | {"category_id":"1"} 
4 | product | 1 | existingcategory1/custom-url-key-for-product-123456789/ | catalog/product/view/id/1/category/1 | 0 | 2 | NULL | 1 | {"category_id":"1"} 
5 | product | 1 | existingcategory2/custom-url-key-for-product-123456789/ | catalog/product/view/id/1/category/2 | 0 | 1 | NULL | 1 | {"category_id":"2"}
6 | product | 1 | existingcategory2/custom-url-key-for-product-123456789/ | catalog/product/view/id/1/category/2 | 0 | 2 | NULL | 1 | {"category_id":"2"} 

3. Import csv with following information (just removed the category information)

sku | url_key | product_websites | store_view_code
------------ | ------------- | ------------- | ------------- 
123456789 | custom-url-key-for-product-123456789 | de,at | 
123456789 | custom-url-key-for-product-123456789-de |  | de
123456789 | custom-url-key-for-product-123456789-at |  | at

4. This also works with following csv
3. Import csv with following information (just removed the category information)

sku | url_key | product_websites
------------ | ------------- | ------------- 
123456789 | custom-url-key-for-product-123456789 | de,at

### Expected result

1. The Url Rewrite Table looks the same as before, because we doesn't change any categories and wanted them inside too!

### Actual result

1. Just the follwing information are inside of the url_rewrite table:

| url_rewrite_id | entity_type | entity_id | request_path | target_path | redirect_type | store_id | description | is_autogenerated | metadata
------------ | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- 
1 | product |  1 | custom-url-key-for-product-123456789/ | catalog/product/view/id/1 | 0 | 1 | NULL | 1 | NULL 
2 | product |  1 | custom-url-key-for-product-123456789/ | catalog/product/view/id/1 | 0 | 2 | NULL | 1 | NULL

# Conclusion:
So if you provide an url key but no categories, Magento deletes the existing category url rewrites.
One missing point: 
- if you forgot to provide `product_websites` in your csv file, no url_rewrite is generated or updated. This behavior may be here by default, but i think it's better to renew all information about if the url_key has changed.

Greetz,
Ulf

**Comment by magento-engcom-team** [comment]
Hi @ulftietze. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ulftietze do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @ulftietze it looks like this is duplicate to your issue -> https://github.com/magento/magento2/issues/18234? maybe just edit referenced issue.

**Comment by ulftietze** [comment]
Hey @engcom-backlog-nazar,

no this is not a duplicate due to the product categorie url rewrites now.
In #18234 we deal with the category url rewrites during creation. This time we have problems with individual product rewrites.
Have a look on my description, this are rewrites with entity_type "product".

I already have a fix for this, but first i have to write/fix some unit test ^^

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @ulftietze thank you for you report, this issue has already fixed in 2.3-develop branch.
